File: android/app/src/main/java/com/desmoines/aipulse/ui/components/SubscriptionTierCard.kt
package com.desmoines.aipulse.ui.components

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.slideInVertically
import androidx.compose.animation.slideOutVertically
import androidx.compose.animation.togetherWith
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.desmoines.aipulse.ui.theme.GlassIntensity
import com.desmoines.aipulse.ui.theme.PremiumTokens
import com.desmoines.aipulse.ui.theme.glassSurface
import com.desmoines.aipulse.util.rememberHapticPerformer
import com.desmoines.aipulse.util.rememberShouldReduceAnimations

/**
 * Value proposition shown on a subscription tier card.
 */
data class TierValueProp(
    val icon: ImageVector,
    val text: String,
)

/**
 * Animated subscription tier card used in the redesigned SubscriptionScreen.
 * Mirrors iOS `SubscriptionTierCard.swift`.
 *
 * The featured tier is scaled up slightly and (for VIP) renders with a gold
 * gradient border. CTA press plays premiumUnlock haptic and scales to 0.97.
 */
@Composable
fun SubscriptionTierCard(
    tierName: String,
    price: String,
    billingLabel: String,
    tagline: String,
    valueProps: List<TierValueProp>,
    ctaLabel: String,
    onCtaClick: () -> Unit,
    modifier: Modifier = Modifier,
    isFeatured: Boolean = false,
    isVip: Boolean = false,
    strikethroughPrice: String? = null,
) {
    val haptic = rememberHapticPerformer()
    val reduceMotion = rememberShouldReduceAnimations()

    val scale by animateFloatAsState(
        targetValue = if (isFeatured) 1.03f else 1.0f,
        animationSpec = spring(dampingRatio = Spring.DampingRatioMediumBouncy),
        label = "card-scale",
    )

    val surface = MaterialTheme.colorScheme.surface

    Box(
        modifier = modifier
            .graphicsLayer {
                scaleX = scale
                scaleY = scale
            }
            .fillMaxWidth()
            .glassSurface(
                shape = RoundedCornerShape(PremiumTokens.CornerXl),
                intensity = GlassIntensity.Card,
                elevation = if (isFeatured) PremiumTokens.Elevation8 else PremiumTokens.Elevation4,
            )
            .then(
                if (isVip) {
                    Modifier.border(
                        width = 2.dp,
                        brush = PremiumTokens.PremiumGoldGradient,
                        shape = RoundedCornerShape(PremiumTokens.CornerXl),
                    )
                } else if (isFeatured) {
                    Modifier.border(
                        width = 1.5.dp,
                        brush = PremiumTokens.BrandGradient,
                        shape = RoundedCornerShape(PremiumTokens.CornerXl),
                    )
                } else Modifier
            )
            .padding(20.dp)
            .semantics {
                contentDescription = buildString {
                    append("$tierName tier. ")
                    append("$price $billingLabel. ")
                    append(tagline)
                    append(". ")
                    valueProps.forEach { append(it.text); append(". ") }
                }
            },
    ) {
        Column {
            // Header row with tier name + optional VIP badge
            Row(verticalAlignment = Alignment.CenterVertically) {
                Text(
                    text = tierName,
                    style = MaterialTheme.typography.titleLarge,
                    fontWeight = FontWeight.Bold,
                )
                if (isVip) {
                    Spacer(Modifier.size(8.dp))
                    VipGoldBadge()
                }
            }
            Spacer(Modifier.height(4.dp))
            Text(
                text = tagline,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )

            Spacer(Modifier.height(14.dp))

            // Price row — animated number roll / crossfade on change
            Row(verticalAlignment = Alignment.Bottom) {
                AnimatedContent(
                    targetState = price,
                    transitionSpec = {
                        if (reduceMotion) {
                            fadeIn() togetherWith fadeOut()
                        } else {
                            (slideInVertically { h -> h } + fadeIn()) togetherWith
                                (slideOutVertically { h -> -h } + fadeOut())
                        }
                    },
                    label = "price-roll",
                ) { p ->
                    Text(
                        text = p,
                        fontSize = 34.sp,
                        fontWeight = FontWeight.ExtraBold,
                    )
                }
                Spacer(Modifier.size(6.dp))
                Text(
                    text = billingLabel,
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(bottom = 6.dp),
                )
                if (strikethroughPrice != null) {
                    Spacer(Modifier.size(8.dp))
                    Text(
                        text = strikethroughPrice,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        textDecoration = TextDecoration.LineThrough,
                        modifier = Modifier.padding(bottom = 6.dp),
                    )
                }
            }

            Spacer(Modifier.height(16.dp))

            // Value props list (top 4)
            valueProps.take(4).forEach { prop ->
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    modifier = Modifier.padding(vertical = 6.dp),
                ) {
                    Box(
                        modifier = Modifier
                            .size(22.dp)
                            .clip(CircleShape)
                            .background(
                                if (isVip) PremiumTokens.PremiumGoldGradient
                                else Brush.linearGradient(
                                    colors = listOf(
                                        MaterialTheme.colorScheme.primaryContainer,
                                        MaterialTheme.colorScheme.primaryContainer,
                                    )
                                )
                            ),
                        contentAlignment = Alignment.Center,
                    ) {
                        Icon(
                            imageVector = prop.icon,
                            contentDescription = null,
                            modifier = Modifier.size(14.dp),
                            tint = if (isVip) Color.White else MaterialTheme.colorScheme.onPrimaryContainer,
                        )
                    }
                    Spacer(Modifier.size(10.dp))
                    Text(
                        text = prop.text,
                        style = MaterialTheme.typography.bodyMedium,
                    )
                }
            }

            Spacer(Modifier.height(18.dp))

            // CTA
            val ctaScale by animateFloatAsState(
                targetValue = 1f,
                animationSpec = spring(dampingRatio = Spring.DampingRatioMediumBouncy),
                label = "cta-scale",
            )
            Button(
                onClick = {
                    haptic.premiumUnlock()
                    onCtaClick()
                },
                modifier = Modifier
                    .fillMaxWidth()
                    .height(52.dp)
                    .graphicsLayer {
                        scaleX = ctaScale
                        scaleY = ctaScale
                    },
                shape = RoundedCornerShape(PremiumTokens.CornerLg),
                colors = ButtonDefaults.buttonColors(
                    containerColor = if (isVip) Color(0xFFB8860B)
                    else MaterialTheme.colorScheme.primary,
                ),
            ) {
                Icon(
                    imageVector = Icons.Filled.Check,
                    contentDescription = null,
                    modifier = Modifier.size(18.dp),
                )
                Spacer(Modifier.size(8.dp))
                Text(
                    text = ctaLabel,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold,
                )
            }
        }
    }
}

/**
 * Monthly / Annual billing toggle with animated thumb.
 * Fires a light haptic on change.
 */
@Composable
fun BillingCycleToggle(
    isAnnual: Boolean,
    onChange: (Boolean) -> Unit,
    modifier: Modifier = Modifier,
) {
    val haptic = rememberHapticPerformer()
    val thumbOffset by animateFloatAsState(
        targetValue = if (isAnnual) 1f else 0f,
        animationSpec = spring(dampingRatio = Spring.DampingRatioMediumBouncy),
        label = "billing-thumb",
    )

    Row(
        modifier = modifier
            .clip(RoundedCornerShape(PremiumTokens.CornerLg))
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .padding(4.dp)
            .semantics {
                contentDescription = if (isAnnual) "Annual billing selected" else "Monthly billing selected"
            },
        horizontalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        @Composable
        fun segment(label: String, selected: Boolean, onClick: () -> Unit) {
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(PremiumTokens.CornerMd))
                    .background(
                        if (selected) MaterialTheme.colorScheme.surface else Color.Transparent
                    )
                    .clickable {
                        haptic.selection()
                        onClick()
                    }
                    .padding(horizontal = 18.dp, vertical = 10.dp),
                contentAlignment = Alignment.Center,
            ) {
                Text(
                    text = label,
                    fontWeight = if (selected) FontWeight.Bold else FontWeight.Normal,
                    color = if (selected) MaterialTheme.colorScheme.primary
                    else MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        segment("Monthly", !isAnnual) { onChange(false) }
        segment("Annual · save 20%", isAnnual) { onChange(true) }
        // thumbOffset is exposed for future slide-thumb implementations
        val _unused = thumbOffset
    }
}

File: ios/DesMoinesInsider/Views/Components/SubscriptionTierCard.swift
import SwiftUI

/// Value prop row shown inside a SubscriptionTierCard.
struct TierValueProp: Identifiable {
    let id = UUID()
    let systemImage: String
    let text: String
}

/// Animated subscription tier card used in the redesigned SubscriptionScreen.
/// Mirrors Android `SubscriptionTierCard.kt`.
///
/// - Featured tier scales up slightly and gets a brand gradient border.
/// - VIP tier gets the premium gold gradient border and badge.
/// - Price crossfades on billing cycle change.
/// - CTA plays `HapticFeedback.shared.premiumUnlock()` on tap.
struct SubscriptionTierCard: View {
    let tierName: String
    let price: String
    let billingLabel: String
    let tagline: String
    let valueProps: [TierValueProp]
    let ctaLabel: String
    var isFeatured: Bool = false
    var isVip: Bool = false
    var strikethroughPrice: String? = nil
    let onCtaTap: () -> Void

    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    // IOS-COMPLY-003: price is critical content — scale the fixed point size with
    // Dynamic Type (relative to .largeTitle) so it grows up to AX5 without clipping.
    @ScaledMetric(relativeTo: .largeTitle) private var priceSize: CGFloat = 34

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            // Header
            HStack(spacing: 8) {
                Text(tierName)
                    .font(.title2.bold())
                if isVip {
                    VipGoldBadge()
                }
                Spacer()
            }

            Text(tagline)
                .font(.subheadline)
                .foregroundStyle(.secondary)
                .padding(.top, 2)

            // Price row with crossfade
            HStack(alignment: .lastTextBaseline, spacing: 6) {
                Text(price)
                    .font(.system(size: priceSize, weight: .heavy))
                    .contentTransition(.numericText())
                    .animation(
                        reduceMotion ? .linear(duration: 0) : .spring(response: 0.4, dampingFraction: 0.75),
                        value: price
                    )
                Text(billingLabel)
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                if let strikethroughPrice {
                    Text(strikethroughPrice)
                        .font(.caption)
                        .foregroundStyle(.secondary)
                        .strikethrough()
                }
            }
            .padding(.top, 14)

            // Value props
            VStack(alignment: .leading, spacing: 10) {
                ForEach(valueProps.prefix(4)) { prop in
                    HStack(spacing: 10) {
                        ZStack {
                            Circle()
                                .fill(isVip
                                      ? AnyShapeStyle(PremiumTokens.premiumGoldGradient)
                                      : AnyShapeStyle(Color.accentColor.opacity(0.18)))
                                .frame(width: 22, height: 22)
                            Image(systemName: prop.systemImage)
                                .font(.system(size: 11, weight: .bold))
                                .foregroundStyle(isVip ? Color.white : Color.accentColor)
                        }
                        Text(prop.text)
                            .font(.subheadline)
                    }
                }
            }
            .padding(.top, 16)

            // CTA
            Button {
                HapticFeedback.shared.premiumUnlock()
                onCtaTap()
            } label: {
                HStack(spacing: 8) {
                    Image(systemName: "checkmark.seal.fill")
                        .font(.system(size: 16, weight: .bold))
                    Text(ctaLabel)
                        .font(.headline)
                }
                .frame(maxWidth: .infinity)
                .frame(height: 52)
                .background(
                    isVip
                        ? AnyShapeStyle(PremiumTokens.premiumGoldGradient)
                        : AnyShapeStyle(Color.accentColor)
                )
                .foregroundStyle(.white)
                .clipShape(RoundedRectangle(cornerRadius: PremiumTokens.cornerLg))
            }
            .buttonStyle(.pressableCard)
            .padding(.top, 18)
        }
        .padding(20)
        .glassCard(cornerRadius: PremiumTokens.cornerXl, material: .regularMaterial, elevation: isFeatured ? PremiumTokens.elevation8 : PremiumTokens.elevation4)
        .overlay(
            RoundedRectangle(cornerRadius: PremiumTokens.cornerXl)
                .strokeBorder(
                    isVip
                        ? AnyShapeStyle(PremiumTokens.premiumGoldGradient)
                        : (isFeatured
                           ? AnyShapeStyle(PremiumTokens.brandGradient)
                           : AnyShapeStyle(Color.clear)),
                    lineWidth: isVip ? 2 : (isFeatured ? 1.5 : 0)
                )
        )
        .scaleEffect(isFeatured ? 1.03 : 1.0)
        .premiumShadow(elevation: isFeatured ? 12 : 4)
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(tierName) tier. \(price) \(billingLabel). \(tagline).")
    }
}

/// Monthly / Annual billing toggle matching Android `BillingCycleToggle`.
///
/// NOTE (IOS-AUDIT-UX-032): this component and `SubscriptionTierCard` are not yet
/// wired to a live screen. Before using them in production, drive `annualLabel`
/// and the card's `price`/`billingLabel` from StoreKit (`Product.displayPrice`
/// and PaywallView's computed `annualSavingsPercent`) — do NOT ship the
/// placeholder "save 20%" / "$14.99" strings, which can diverge from the real
/// products.
struct BillingCycleToggle: View {
    @Binding var isAnnual: Bool
    /// Defaults to a neutral label so a real savings percent (from StoreKit) is
    /// supplied at the call site rather than hardcoded.
    var annualLabel: String = "Annual"

    var body: some View {
        HStack(spacing: 4) {
            segment(label: "Monthly", selected: !isAnnual) { isAnnual = false }
            segment(label: annualLabel, selected: isAnnual) { isAnnual = true }
        }
        .padding(4)
        .background(Color(.secondarySystemBackground))
        .clipShape(RoundedRectangle(cornerRadius: PremiumTokens.cornerLg))
    }

    @ViewBuilder
    private func segment(label: String, selected: Bool, action: @escaping () -> Void) -> some View {
        Button {
            HapticFeedback.shared.selection()
            action()
        } label: {
            Text(label)
                .font(.subheadline.weight(selected ? .bold : .regular))
                .padding(.horizontal, 18)
                .padding(.vertical, 10)
                .background(selected ? Color(.systemBackground) : Color.clear)
                .foregroundStyle(selected ? Color.accentColor : Color.secondary)
                .clipShape(RoundedRectangle(cornerRadius: PremiumTokens.cornerMd))
        }
        .buttonStyle(.plain)
    }
}

#Preview {
    VStack(spacing: 16) {
        BillingCycleToggle(isAnnual: .constant(true))
        SubscriptionTierCard(
            tierName: "VIP",
            price: "$14.99",
            billingLabel: "/ month",
            tagline: "Everything unlocked, zero ads.",
            valueProps: [
                TierValueProp(systemImage: "sparkles", text: "Unlimited favorites"),
                TierValueProp(systemImage: "bell.badge.fill", text: "Early event alerts"),
                TierValueProp(systemImage: "brain.head.profile", text: "AI trip planning"),
                TierValueProp(systemImage: "gift.fill", text: "Exclusive perks")
            ],
            ctaLabel: "Go VIP",
            isFeatured: true,
            isVip: true,
            onCtaTap: {}
        )
    }
    .padding()
}
